[#136] Add 'recent activities' to tracking screen

elbenfreund opened the pull request:
This adds a 'recent activities' area to the tracking screen that can be used to take previous activities as templates for new raw fact strings or start tracking them right away. The entire functionality is optional (on by default right now) and can be configures via preferences.

The 'recent activities' box size can be adjusted by specifying the amount of activities it shall be able to show. All activities that not fit are available via scrolling.

Closes: #136

elbenfreund committed:
Add inline comments

elbenfreund committed:
Add dedicated helper for fetching recent activities

elbenfreund committed:
Add 'recent activities' to start-tracking view

This commits extends the 'start tracking' view with a list of recent
activities that can be used to euther copy the data to the entry field,
ready to be extended and then started manually by the user or to
start/continue tracking this activity right away.

Ref: #136

elbenfreund committed:
Sizing for recent activity widget

elbenfreund committed:
Adjust existing tests

elbenfreund committed:
Wrap 'make test' with 'xvfb-run'

Closes: 206

jtojnar commented on hamster_gtk/tracking/screens.py:
```diff
@@ -238,3 +268,119 @@ def _on_start_tracking_button(self, button):
+    def _on_config_changed(self, sender):
+        """Callback triggered when 'config-changed' event fired."""
+        if self._app.config['tracking_show_recent_activities']:
+            # We re-create it even if one existed before necause its parameters
```
necause

jtojnar commented on hamster_gtk/preferences/preferences_dialog.py:
```diff
@@ -80,6 +80,11 @@ def __init__(self, parent, app, initial, *args, **kwargs):
                  (_("Autocomplete activities and categories separately"),
                   HamsterSwitch())),
+                ('tracking_show_recent_activities',
+                 (_("Allow tracking based on recent activities."),
```
This does not sound clear to me. I would choose something like “Show recent activities for quickly starting tracking”

jtojnar commented on hamster_gtk/helpers.py:
```diff
@@ -177,6 +180,23 @@ def decompose_raw_fact_string(text, raw=False):
+    return OrderedSet(recent_activities)
+
+
+def serialize_activity(activity):
```
Do not we have something like this somewhere already?

elbenfreund replied:
I think earlier ``hamster-lib`` branches had something like this. Also the overview has its own "serialize activity" code which is different however:
- it deals with legacy hamerest representation of ``activity.category=None``. This is something that I would be willing to throw out of the window.
- It uses `` - `` as seperator between ``activity.name`` and ``category``. I think this is better looking for the overview. What do you think?

So this helper provides a generic serialization utilizing the standart ``@`` seperator. As far as I can tell current versions of ``hamster-lib`` or ``hamster-gtk`` do not have such a function yet. But share the feeling that we had this around for a while in various incarnations...

jtojnar replied:
* Wait, do we intend to not support category-less facts?
* I actually prefer the at sign for consistency or habit.

elbenfreund replied:
- No we do and want to support category less facts. I am not quite sure that the way legacy hamster did is really where we want to go.. 
- Does that mean you would prefer ``@`` in the overview as well? At this stage I do not care about this detail very much as we will have to revisit the general layout at some later stage anyway...

jtojnar replied:
Yes, having `@` in overview is what I am accustomed to.

elbenfreund replied:
Alrighty. Will adjust this then :) Im cool either way tbh.

jtojnar commented on hamster_gtk/hamster_gtk.py:
```diff
@@ -279,6 +285,8 @@ def _get_default_config(self):
             'autocomplete_activities_range': 30,
             'autocomplete_split_activity': False,
+            'tracking_show_recent_activities': True,
+            'tracking_recent_activities_items': 6,
```
I would use `count` or `number`. `items` is less clear.

jtojnar commented on hamster_gtk/tracking/screens.py:
```diff
@@ -230,6 +239,27 @@ def reset(self):
+        # allocated to its children so they actually have a height that we can
+        # use.
+        grid.show_all()
+        # We fetch an arbitrary Button as height-reference
```
this is really ugly and crashes when there are no items:

```
$ env XDG_CONFIG_HOME=$HOME/.dev_config XDG_DATA_HOME=$HOME/.dev_local hamster-gtk
Hamster-GTK started.
Traceback (most recent call last):
  File "/nix/store/fzrxvdjas3icqb2n5akahvix758jnfvj-python3.6-hamster-gtk/lib/python3.6/site-packages/hamster_gtk/hamster_gtk.py", line 213, in _activate
    self.window = MainWindow(app)
  File "/nix/store/fzrxvdjas3icqb2n5akahvix758jnfvj-python3.6-hamster-gtk/lib/python3.6/site-packages/hamster_gtk/hamster_gtk.py", line 107, in __init__
    self.add(TrackingScreen(self.app))
  File "/nix/store/fzrxvdjas3icqb2n5akahvix758jnfvj-python3.6-hamster-gtk/lib/python3.6/site-packages/hamster_gtk/tracking/screens.py", line 49, in __init__
    self.start_tracking_view = StartTrackingBox(self._app)
  File "/nix/store/fzrxvdjas3icqb2n5akahvix758jnfvj-python3.6-hamster-gtk/lib/python3.6/site-packages/hamster_gtk/tracking/screens.py", line 195, in __init__
    self.recent_activities_widget = self._get_recent_activities_widget()
  File "/nix/store/fzrxvdjas3icqb2n5akahvix758jnfvj-python3.6-hamster-gtk/lib/python3.6/site-packages/hamster_gtk/tracking/screens.py", line 255, in _get_recent_activities_widget
    child = grid.get_children()[1]
IndexError: list index out of range
Traceback (most recent call last):
  File "/nix/store/fzrxvdjas3icqb2n5akahvix758jnfvj-python3.6-hamster-gtk/lib/python3.6/site-packages/hamster_gtk/hamster_gtk.py", line 218, in _activate
    app.add_window(self.window)
TypeError: Argument 1 does not allow None as a value
Hamster-GTK shut down.
```

elbenfreund replied:
I agree that the way we derive our min_height hint is far from elegant (but fixed value layouts are far worse even). I am absolutely open to better approaches! Empty children are definitely a nice catch and have been addressed in the latest version.

jtojnar replied:
Maybe add a todo comment and go with this.

elbenfreund replied:
I created #224  as I would like to avoid todo comments.

jtojnar commented on hamster_gtk/tracking/screens.py:
```diff
@@ -238,3 +268,119 @@ def _on_start_tracking_button(self, button):
+
+    def _populate(self):
+        """Fill the widget with rows per activity."""
+        def add_row_widgets(row_counter, activity):
```
`row_counter` → `row_index`/`row_number`; counter implies an object that counts

jtojnar commented on hamster_gtk/helpers.py:
```diff
@@ -177,6 +180,23 @@ def decompose_raw_fact_string(text, raw=False):
+def get_recent_activities(controller, start, end):
+    """Return a list of all activities logged in facts within the given timeframe."""
+    recent_activities = [fact.activity for fact in controller.facts.get_all(start=start, end=end)]
+    return OrderedSet(recent_activities)
```
The activities are returned in a convoluted order. It should be reversed prior to removing duplicates.

elbenfreund replied:
I am not sure I do understand. Could you elaborate please?

jtojnar replied:
It should be `OrderedSet(reversed(recent_activities))`, otherwise you will get the activities in the order they were first used, not the last.

```
facts     1 2 1 3 2 1 3  output   reversed
current   x x   x       → 1 2 3  → 3 2 1
expected          x x x → 3 1 2  → 2 1 3
```

elbenfreund replied:
Thank you for the clarification. To be honest, right now ``FactManager.get_all()`` does not commit to any ordering at all. I created an [issue with ``hamster-lib``](https://projecthamster.atlassian.net/browse/LIB-259) to improve this.
For now I will implement a workaround simple sorting call as part of the prelimery ``get_recent_activities`` helper which will hopefully address your concerns.

jtojnar commented on hamster_gtk/tracking/screens.py:
```diff
@@ -238,3 +268,119 @@ def _on_start_tracking_button(self, button):
+
+        today = datetime.date.today()
+        start = today - datetime.timedelta(90)
+        activities = reversed(helpers.get_recent_activities(self._controller, start, today))
```
Do we really want all the activities and scroll instead of displaying the `tracking_recent_activities_items`?

elbenfreund replied:
I would prefer such a more general solution. There is nothing preventing us from limiting the results to just ``tracking_recent_activities_items`` (via config) if we want, but especially given that we already have a solution I do not see the harm. 

We **need** scrolling anyway (because we can not make any guarantee that ``tracking_recent_activities_items`` will fit in the space available. In light of that, it seems we get the extra activities for free if the user bothers scolling.

What would be the downside?

jtojnar replied:
Accidental scrolling to the other end and performance.

elbenfreund replied:
- "accidental scrolling": Im not sure I see the use case. especially as there is the posibility of a scrollbox even if if we just fetch the number of activities defined in the config.
- by default we just fetch todays facts. this should net us no more than a couple of dozens facts at most I figure. I doubt there is any concievable performance impact on the DB or the GUI rendering. Unless I'm missing something this seems like a case of premature optimization.

jtojnar replied:
* You scroll somewhere accidentally and since the list is long you do not have a simple way of returning back – you cannot just quickly scroll down and when you find out that you are on the wrong side quickly scroll to the correct one.
* I see more than just today’s facts. In fact there does not seem to be any limit.

elbenfreund replied:
You are [right](https://github.com/projecthamster/hamster-gtk/blob/feature/136/hamster_gtk/tracking/screens.py#L354)!

The idea was to make the timedelta adjustable via a config setting, but i forgot to actually do that :(
If done that way it is up to the user to figure what timeframe is viable to them. We probably should point out the performance consideration though... 

It would probably be best to leave this additional customization to a seperate issue and use ``start=today`` for now...

elbenfreund committed:
rename 'row_counter' parameter to 'row_index'

jtojnar commented on hamster_gtk/tracking/screens.py:
```diff
@@ -45,7 +45,8 @@ def __init__(self, app, *args, **kwargs):
         self.set_transition_duration(1000)
         self.current_fact_view = CurrentFactBox(self._app.controller)
-        self.current_fact_view.connect('tracking-stopped', self.update)
+        self.current_fact_view.con275G
```
wah

elbenfreund committed:
Fix typo

elbenfreund committed:
Change wording within preferences.

elbenfreund committed:
fix ordering for 'get_recent_activities' helper

elbenfreund committed:
Account for empty children when figuring out min_height

elbenfreund committed:
Recent activities box uses last 24h

elbenfreund committed:
Fix docstring

elbenfreund committed:
Improve 'config' fixture

Until now, we returned a ``PathLocal`` instance as value for the
``tmpfile_path`` key. This has been changed to a proper string
representations instead.

elbenfreund committed:
Fix 'app' fixture

Our ``app`` fixture (which is used extensively) used to instantiate the
default ``HamsterGTK`` class. This meant that the default config loading
was triggered and as a consequence the systems live config being used.
This is obviously an issue. Instead we monkey patch the class now to
skip the config loading machinery and use/return the ``config`` fixture.

elbenfreund committed:
Fix test if 'HamsterGTK' instantiation

Our previous test simply instantiated the class, hence tiggering the
config loading/creation machinery. This will directly interfere with the
actual user config! To avoid this we provide a new test that
monkeypatches the class accordingly. As a consequence the config loading
needs to be tested seperatly now though.

jtojnar commented on tests/test_hamster-gtk.py:
```diff
@@ -14,9 +14,19 @@
-        app = hamster_gtk.HamsterGTK()
+    def test_instantiation(self, config):
+        """
+        Make sure class instatiation works as intended.
```
instatiation → instantiation

elbenfreund committed:
Rename 'tracking_recent_activities_items' to 'count'

elbenfreund committed:
helpers: Refactor 'serialize_activity'

Ref.: #136

elbenfreund commented:
@jtojnar I addressed most of your concerns and CI is passing again (finally). Do you think you will find some time to review it again?

The only big issue left open is [this](https://github.com/projecthamster/hamster-gtk/pull/208#discussion_r139186084). I am absolutely open to improve the current hackish solution if you have better idea.

jtojnar commented on hamster_gtk/helpers.py:
```diff
@@ -196,10 +197,34 @@ def get_recent_activities(controller, start, end):
+        separator (str, optional): ``string`` used to separate ``activity.name`` and
+            ``category.name``. The separator will be omitted if ``none_categoty=''`` and
+            ``activity.category=None``. Defaults to ``@``.
+        none_category (str, optional): ``string`` to represent the 'lack of a cactegory' for an
```
`cactegory`

jtojnar commented on hamster_gtk/helpers.py:
```diff
@@ -177,6 +182,54 @@ def decompose_raw_fact_string(text, raw=False):
+def get_recent_activities(controller, start, end):
+    """Return a list of all activities logged in facts within the given timeframe."""
+    # [FIXME]
+    # This manual sorting within python is of cause less than optimal. We stick
```
of cause → of course

jtojnar commented on hamster_gtk/helpers.py:
```diff
@@ -177,6 +182,54 @@ def decompose_raw_fact_string(text, raw=False):
+    Args:
+        activity (Activity): ``Activity`` instance to serialize.
+        separator (str, optional): ``string`` used to separate ``activity.name`` and
+            ``category.name``. The separator will be omitted if ``none_categoty=''`` and
```
none_categoty → none_category

jtojnar commented on tests/test_hamster-gtk.py:
```diff
@@ -14,9 +14,19 @@
+
+        We actually test against a monkeypatched class in order to avoid the
+        config loading machinery as this would access the user data on fs.
+        The relevant skiped methods need to be tested separately.
```
skipped

jtojnar reviewed (approved):
Several typos, otherwise looks good from a shallow review (ignoring my bloat concerns)

jtojnar commented on hamster_gtk/helpers.py:
```diff
@@ -177,6 +182,54 @@ def decompose_raw_fact_string(text, raw=False):
+    return OrderedSet(recent_activities)
+
+
+def serialize_activity(activity, separator='@', none_category='not categorized'):
```
We probably do not want to serialize uncategorized activities with the “not categorized” category.

![screenshot from 2018-05-04 14-40-33](https://user-images.githubusercontent.com/705123/39628421-53d9893e-4fa9-11e8-8c16-45e3193de61a.png)

If you copy/start them, the fake category will become real.

elbenfreund replied:
good catch. :+1:

elbenfreund committed:
tracking: Improve categoryless fact serialization

When serializing facts for the 'continue tracking' functionality, we
need to make sure that the 'uncategorized' placeholder is omitted,
otherwise a category of that name would be created!

Ref.: #136

elbenfreund committed:
Fix typos

elbenfreund committed:
tracking: Add issue reference to ToDo comment

Ref,: #136

elbenfreund committed:
helpers: Remove 'none_category' argument on serialize_activity

elbenfreund committed:
tests: Fix docker setup

jtojnar commented on hamster_gtk/tracking/screens.py:
```diff
@@ -238,3 +270,119 @@ def _on_start_tracking_button(self, button):
+            """
+            def get_label(activity):
+                """Label representing the activity/category combination."""
+                label = Gtk.Label(helpers.serialize_activity(activity, none_category=''))
```
You will still need to remove `none_category` from here.

elbenfreund committed:
Fix imports

elbenfreund committed:
helpers: Add tests for 'serialize_activity'
